File: go/yo/yo.go
package main

import (
	"fmt"
	"code.google.com/p/gcfg"
	"io/ioutil"
	"net/http"
	"net/url"
	"os"
	"path"
)

const Endpoint = "http://doneyo.herokuapp.com"
const ConfigFileName = ".yorc"


func data(user string) url.Values {
	return url.Values{"user": {user}}
}

func initialConfig(user string) string {
	return fmt.Sprintf("[user]\nname = %s", user)
}

func yo(user string) {
	response, error := http.PostForm(Endpoint, data(user))

	if error != nil { os.Exit(1) }

	message, _ := ioutil.ReadAll(response.Body)
	if response.StatusCode == 200 {
		fmt.Println(string(message))
	} else {
		fmt.Fprintln(os.Stderr, string(message))
		os.Exit(1)
	}
}

type Config struct {
	User struct {
		Name string
	}
}

func ConfigFile() string {
	return path.Join(os.Getenv("HOME"), ConfigFileName)
}

func username() string {
	if _, error := os.Stat(ConfigFile()); error != nil {
		askUser()
	}

	var config Config
	error := gcfg.ReadFileInto(&config, ConfigFile())
	if error != nil { panic(error) }

	return config.User.Name
}

func askUser() {
	var user string

	fmt.Printf("Your Yo username? : ")
	fmt.Scanf("%s", &user)
	createConfigFile(user)
}

func createConfigFile(user string) {
	error := ioutil.WriteFile(ConfigFile(), []byte(initialConfig(user)), 0644)
	if error != nil { panic(error) }
}


func main() {
	yo(username())
}

File: lib/doneyo/yo.rb
require 'net/http'

module Doneyo
  class Yo
    ENDPOINT = 'http://doneyo.herokuapp.com'
    CONFIG_FILE_NAME = '.yorc'

    class << self
      def send
        begin
          puts new(username).send
        rescue
          $stderr.puts $!.to_s
          abort
        end
      end


      private

      def config_file
        File.join(ENV['HOME'], CONFIG_FILE_NAME)
      end

      def username
        ask_user unless File.exists?(config_file)

        hash = parse_into_hash(File.read(config_file))
        hash['[user]'] =~ /name\s*=\s(\S+)/
        $1
      end

      def ask_user
        print 'Your Yo username? : '
        create_config_file gets
      end

      def create_config_file(user)
        user = user.to_s.strip
        raise 'Username required' if user == ''

        open(config_file, 'w') do |f|
          f.write "[user]\nname = #{user}"
        end
      end

      def parse_into_hash(string)
        Hash[
          string.split(/(\[[^\[\]]+\])/)[1..-1].each_slice(2).to_a
        ]
      end
    end


    def initialize(user)
      @user = user
    end

    def send
      response = Net::HTTP.post_form(URI.parse(ENDPOINT), user: @user)
      response.body
    end
  end
end
